File: ROID-Forum-Server/ROID-Forum-Server/Convert Server Code/GMGServer/bin/Server/LoginController.dart
library LoginController;
import 'Server.dart';
import 'User.dart';
import 'dart:convert';

class LoginController {
  Server server;
  LoginController(Server s) {
    server = s;
  }
  void logic() {
  }
  void onMessage(User u, Map message) {
    if (message['Title'] == 'Set Avatar' && u.account != null) {
      u.account.avatarURL = message['AvatarURL'];
    }
    if (message['Title'] == 'Get Avatar' && u.account != null) {
      Map m = new Map();
              m['Controller'] = 'Login';
              m['Title'] = 'Get Avatar';
              m['AvatarURL'] =  u.account.avatarURL ;
              String s = JSON.encode(m);
              u.sendString(s);
        }
    if (message['Title'] == 'Login' && u.account == null) {
      if (server.accountController.accountExists(message['Name'], message['Password'])) {
        u.account = server.accountController.getAccount(message['Name'], message['Password']);
        //Send login notification
        Map m = new Map();
        m['Controller'] = 'Login';
        m['Title'] = 'Logged In';
        m['Name'] = u.account.name;
        m['Password'] = u.account.password;
        String s = JSON.encode(m);
        u.sendString(s);

        server.accountLoggedIn(u);
      } else {
        //Send login failure
        Map m = new Map();
        m['Controller'] = 'Login';
        m['Title'] = 'Login Failed';
        String s = JSON.encode(m);
        u.sendString(s);
      }
    }
    if (message['Title'] == 'Logout' && u.account != null) {
      //Perhaps an account method can be called for saving or other logic
      u.account = null;
      //Send logout notification
      Map m = new Map();
      m['Controller'] = 'Login';
      m['Title'] = 'Logged Out';
      String s = JSON.encode(m);
      u.sendString(s);
      server.accountLoggedOut(u);
    }
    if (message['Title'] == 'Register' && u.account == null) {
      if (server.accountController.accountNameExists(message['Name'])) {
        //Send register failure
        Map m = new Map();
        m['Controller'] = 'Login';
        m['Title'] = 'Register Failed';
        String s = JSON.encode(m);
        u.sendString(s);
      } else {
        u.account = server.accountController.createAccount(message['Name'], message['Password'], message['Email']);
        //Send login notification
        Map m = new Map();
        m['Controller'] = 'Login';
        m['Title'] = 'Logged In';
        m['Name'] = u.account.name;
        m['Password'] = u.account.password;
        String s = JSON.encode(m);
        u.sendString(s);
        server.accountLoggedIn(u);
      }
    }
  }
}

File: ROID-Forum-Server/ROID-Forum-Server/LoginController.cs
using System;
using System.Collections.Generic;
using Cassandra;
namespace ROIDForumServer
{
    public class LoginController (ServerController serverController)
    {
        private ServerController Server { get; } = serverController;
        private ISession DatabaseSession { get; } = serverController.Database.GetSession();
        public void OnMessage(ConnectedUser user, Dictionary<string, object> message)
        {
            if ((string)message["Title"] == "Set Avatar" && user.AccountId != null)
            {
                DatabaseAccount.SetAvatarUrl(DatabaseSession, (Guid)user.AccountId, (string)message["AvatarURL"]);
            }
            if ((string)message["Title"] == "Get Avatar" && user.AccountId != null)
            {
                user.Send(ServerMessages.GetAvatarMessage(DatabaseAccount.GetAvatarUrl(DatabaseSession, (Guid)user.AccountId)));
            }
            if ((string)message["Title"] == "Login" && user.AccountId == null)
            {
                string username = (string)message["Name"];
                string password = (string)message["Password"];
                if (String.IsNullOrWhiteSpace(username) || String.IsNullOrWhiteSpace(password))
                {
                    user.Send(ServerMessages.LoginFailedMessage());
                }
                else
                {
                    Guid? accountId = DatabaseAccount.GetAccountIdForCredentials(DatabaseSession, username, password);
                    if (accountId == null)
                    {
                        user.Send(ServerMessages.LoginFailedMessage());
                    }
                    else
                    {
                        user.AccountId = accountId;
                        user.Send(ServerMessages.LoggedInMessage(DatabaseAccount.GetAccountName(DatabaseSession, (Guid)user.AccountId)));
                        Server.AccountLoggedIn(user);
                    }
                }
            }
            if ((string)message["Title"] == "Logout" && user.AccountId != null)
            {
                //Perhaps an account method can be called for saving or other logic
                user.AccountId = null;
                //Send logout notification
                user.Send(ServerMessages.LoggedOutMessage());
                Server.AccountLoggedOut(user);
            }
            if ((string)message["Title"] == "Register" && user.AccountId == null)
            {
                string username = (string)message["Name"];
                string password = (string)message["Password"];
                string email = (string)message["Email"];
                if (String.IsNullOrWhiteSpace(username) || String.IsNullOrWhiteSpace(password) || String.IsNullOrWhiteSpace(email))
                {
                    user.Send(ServerMessages.RegisterFailedMessage());
                    return;
                }
                var (createAccountStatus, accountId) = DatabaseAccount.CreateAccount(DatabaseSession, username, password, email);
                if (createAccountStatus == DatabaseAccount.CreateAccountStatus.AlreadyExists || accountId == null)
                {
                    user.Send(ServerMessages.RegisterFailedMessage());
                    return;
                }
                // Set the account ID to log them in
                user.AccountId = accountId;
                //Send login notification
                user.Send(ServerMessages.LoggedInMessage(DatabaseAccount.GetAccountName(DatabaseSession, (Guid)user.AccountId)));
                Server.AccountLoggedIn(user);
            }
        }
    }
}
